### PR TITLE
Allow layers control section to resize to contents

### LINKED ISSUE #2337: Text overflow in Labels Layer on Linux
## 🐛 Bug
The text for some of the options for the labels layer is not displaying correctly:
image

This persists even with window resizing.

## To Reproduce
I just run the layers.py python example and added a labels layer by clicking the button to add one.

## Expected behavior
Text looks good.
## Environment
```
napari: 0.4.7.dev4+geaa5474
Platform: Linux-5.8.0-7630-generic-x86_64-with-glibc2.31
System: Pop!_OS 20.04 LTS
Python: 3.9.1 | packaged by conda-forge | (default, Jan 26 2021, 01:34:10) [GCC 9.3.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.19.5
SciPy: 1.6.0
Dask: 2021.01.1
VisPy: 0.6.6

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 20.2.6
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- svg: 0.1.4
```
 - Please copy and paste the information at napari info option in help menubar here:

 - Any other relevant information:

### LINKED ISSUE #5472: Visual Bug: Labels Layer Controls get squished when toggling 3D
## 🐛 Bug

When toggling to 3D, the labels layer Layer Controls widget gains an extra line `rendering`.
However the widget doesn't resize for this, so it results in a visual bug of everything squished and partially cut off:
image

If you manually expand the widget, it will look fine. However, in contrast to the 2D version of the widget, it will let you resize it vertically to be too small, squishing the contents again.

## To Reproduce

1. open napari
2. make a labels layer (can be empty)
3. toggle 3D

## Expected behavior

The widget should either resize to permit the extra line item or start out sufficiently large that when the line item is added the visual isn't squished.

## Environment

macOS 13.1, pyqt5, 0.4.17

### PR DESCRIPTION
**Pull request by dalthviz**
# Description

Set `min-height` and `min-width` to the empty widget used by the layers controls container widget instead of the widget itself to allow it to resize dependending on the controls being shown

## Preview

resize2

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

closes #2337 
closes #5472 

# How has this been tested?

- [x] all tests pass with my change
- [x] I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality

### CODE REVIEW

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 1]
```diff
@@ -267,12 +267,15 @@ QLabel#logo_silhouette {
 
 /* ------------------------------------------------------ */
 
+QFrame#empty_controls_widget {
+    min-height: 295px;
```
Comment (psobolewskiPhD) [inline napari/_qt/qt_resources/styles/02_custom.qss]:
```suggestion
    min-height: 225px;
```
What do you think of making the empty height based of the Shapes layer which has the smallest Layer Controls widget?
Then every other one resizes the widget larger.
Note: they don't shrink back, once resized larger that becomes the new min, but that's fine? 🤷‍♂️

**Review by Czaki — APPROVED** [review]

**Review by psobolewskiPhD — APPROVED** [review]
😍 
Outstanding, so clever! I couldn't figure out how to make it work, was going to make PR to just increase the min-height to 325.
I love it, works great with all the layers I've tested (points, shapes, labels, image, surface)

File: napari/_qt/qt_resources/styles/02_custom.qss (hunk 1 above)
Comment (psobolewskiPhD) [reply napari/_qt/qt_resources/styles/02_custom.qss]: Actually Image is the smallest, but then it's like 180.

Comment (dalthviz) [reply napari/_qt/qt_resources/styles/02_custom.qss]: > What do you think of making the empty height based of the Shapes layer which has the smallest Layer Controls widget?

I'm not totally sure from where the min values come from (I guess at some point those looked good in a standard setup on MacOS running a specific example code or for the most used type of layer?)  but decreasing it a little bit shouldn't be a problem with the changes here so I will apply the suggestion :+1:

> Note: they don't shrink back, once resized larger that becomes the new min, but that's fine? 🤷‍♂️

Maybe calling `adjustSize` when setting the current widget could help, will test :+1:

Comment (psobolewskiPhD) [reply napari/_qt/qt_resources/styles/02_custom.qss]: The old min was exactly labels layer controls in 2D on my Mac. Labels has the biggest widget.

Comment (dalthviz) [reply napari/_qt/qt_resources/styles/02_custom.qss]: Did a quick check with `adjustSize` but seems like that doesn't work well (probably needs also a call from the dockwidget) 🤔 If there is no problem with resize large becoming new min then the changes here should be okay

Comment (psobolewskiPhD) [reply napari/_qt/qt_resources/styles/02_custom.qss]: Yup! An improvement for sure!
You're a wizard! 🧙‍♂️

**Comment by psobolewskiPhD** [comment]
Not sure what's up with Windows CI, but I've re-run it twice and it's the same, so I think it's not random?
https://github.com/napari/napari/actions/runs/3905492648/jobs/6674079073#step:8:221
```
napari\_qt\_tests\test_qt_viewer.py .........................ss......... [ 80%]
  ss...Windows fatal exception: access violation
```

**Review by brisvag — APPROVED** [review]
Awesome! This seems like the "correct" solution :)

> Not sure what's up with Windows CI, but I've re-run it twice and it's the same, so I think it's not random?

This looks familiar... I was stuck due to a bunch of these failures in #5312, and I had to ultimately disable some tests on windows CI. Not sure if it's the same here...

**Comment by psobolewskiPhD** [comment]
I tried bumping windows CI to python 3.9 (maybe we should do that anyways, since that's the version we suggest in installation?) but it didn't help.
https://github.com/psobolewskiPhD/napari/actions/runs/3919686014/jobs/6700847765

Can we assume from the 3 dots after skip:
image

That the last 3 tests completed, so it's the first of the next test file?
I'ma try that that in my branch.

**Comment by Czaki** [comment]
could you try to run a verbose test (to find failing job).

It will be also nice to try to run only `napari\_qt\tests\test_qt_viewer.py` to check if fils are related only to these tests.

**Comment by psobolewskiPhD** [comment]
I got pyqt to pass with +1 skip `def test_surface_mixed_dim(make_napari_viewer):`
https://github.com/psobolewskiPhD/napari/actions/runs/3920033829/jobs/6701406395
But now pyside is still cranky 😡 

Edit: also I was not reading the `s` and dots right 🤣 

Edit: it's like whack-a-mole, the viewer tests pass (with skips) but now octree fails:
https://github.com/psobolewskiPhD/napari/actions/runs/3920145076/jobs/6701589988
😡

**Comment by psobolewskiPhD** [comment]
Ok, I ended up having to skip 3 tests plus the octree.
https://github.com/psobolewskiPhD/napari/actions/runs/3920450156/jobs/6702331229

The octree one fails and gives a similar error from the subprocess, so I think it's related, but 🤷‍♂️

**Comment by dalthviz** [comment]
Should I try checking if reverting the min-height change in the qss file helps with the failing checks?

**Comment by psobolewskiPhD** [comment]
> Should I try checking if reverting the min-height change in the qss file helps with the failing checks?

I guess? hard to imagine that being the culprit, but your commit before mu suggestion seems to have flown by. However, my fork branch is using the 295 value and still shows the same CI issues.
Edit: Wait, my branch is py39, I will revert to 3.8 🤷‍♂️ 
Edit2: Reverted, commented out the skips and we're back to crashing on CI:
https://github.com/psobolewskiPhD/napari/actions/runs/3940733794/jobs/6742228661
It is curious that it somehow passed 5 days ago. maybe something changed on CI? 🤷‍♂️

**Comment by brisvag** [comment]
We also have some flaky tests that w're slowly gettin grid of, so some failures are unfortunately fake :/

**Comment by dalthviz** [comment]
Note: seems like the skipped tests are failing due to an access violation error on Windows when Vispy code is creating/managing the OpenGL related context/objects?

Some tracebacks/info from the errors/failures on skipped tests:

* napari\layers\image\experimental\_tests\test_octree_import.py:
```
  stdout = b'\xe2\x94\x80\xe2\x94\x80\xe2\x94\x80\xe2\x94\x80\xe2\x94\x80\xe2\x94\x80\xe2\x94\x80\xe2\x94\x80\xe2\x94\x80\xe2\x94... _make_viewer_then\r\n  "napari\\view_layers.py", line 179, in view_image\r\n  "<string>", line 4, in <module>\r\n\r\n'
  stderr = b'',
  retcode = 3221225477
```
* napari\napari\napari\_qt\_tests\test_qt_viewer.py
```
 Current thread 0x00000e48 (most recent call first):
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\gl\_gl2.py", line 414 in glDrawArrays
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\glir.py", line 1357 in draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\glir.py", line 786 in _parse
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\glir.py", line 824 in parse
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\glir.py", line 506 in flush
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\glir.py", line 584 in flush
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\context.py", line 172 in flush_commands
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\gloo\program.py", line 543 in draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\visuals\shaders\program.py", line 102 in draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\visuals\visual.py", line 451 in draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\scene\visuals.py", line 103 in draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\visuals\visual.py", line 605 in draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\scene\visuals.py", line 103 in draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\scene\canvas.py", line 315 in draw_visual
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\scene\canvas.py", line 277 in _draw_scene
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\scene\canvas.py", line 218 in on_draw
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\util\event.py", line 469 in _invoke_callback
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\util\event.py", line 453 in __call__
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\app\backends\_qt.py", line 903 in paintGL
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\app\backends\_qt.py", line 567 in event
    File "D:\tmp\.tox\py38-windows-pyqt5\lib\site-packages\vispy\app\backends\_qt.py", line 567 in event
    File "D:\a\napari\napari\napari\_qt\layer_controls\qt_layer_controls_container.py", line 126 in _display
    File "D:\a\napari\napari\napari\utils\events\event.py", line 783 in _invoke_callback
    File "D:\a\napari\napari\napari\utils\events\event.py", line 757 in __call__
    File "D:\a\napari\napari\napari\utils\events\containers\_selection.py", line 104 in active
    File "D:\a\napari\napari\napari\utils\events\containers\_selectable_list.py", line 69 in insert
    File "D:\a\napari\napari\napari\components\layerlist.py", line 167 in insert
    File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\_collections_abc.py", line 962 in append
    File "D:\a\napari\napari\napari\components\viewer_model.py", line 826 in add_image
    File "D:\a\napari\napari\napari\utils\migrations.py", line 44 in _update_from_dict
    File "D:\a\napari\napari\napari\_qt\_tests\test_qt_viewer.py", line 522 in test_leaks_image
```
* napari\_qt\_tests\test_qt_notifications.py
```
  Current thread 0x00000c58 (most recent call first):
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\backends\_qt.py", line 823 in _init_specific
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\backends\_qt.py", line 372 in __init__
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\canvas.py", line 228 in create_native
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\canvas.py", line 211 in __init__
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\scene\canvas.py", line 135 in __init__
    File "D:\a\napari\napari\napari\_vispy\canvas.py", line 33 in __init__
    File "D:\a\napari\napari\napari\_qt\qt_viewer.py", line 410 in _create_canvas
    File "D:\a\napari\napari\napari\_qt\qt_viewer.py", line 224 in __init__
    File "D:\a\napari\napari\napari\_qt\qt_main_window.py", line 112 in __init__
    File "D:\a\napari\napari\napari\_qt\qt_main_window.py", line 565 in __init__
    File "D:\a\napari\napari\napari\viewer.py", line 67 in __init__
    File "D:\a\napari\napari\napari\utils\_testsupport.py", line 242 in actual_factory
    File "D:\a\napari\napari\napari\_qt\_tests\test_qt_notifications.py", line 280 in test_notifications_error_with_threading
```
*  napari\_qt\_tests\test_qt_provide_theme.py
```
Current thread 0x000016a8 (most recent call first):
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\backends\_qt.py", line 823 in _init_specific
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\backends\_qt.py", line 372 in __init__
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\canvas.py", line 228 in create_native
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\app\canvas.py", line 211 in __init__
    File "D:\tmp\.tox\py38-windows-pyside2\lib\site-packages\vispy\scene\canvas.py", line 135 in __init__
    File "D:\a\napari\napari\napari\_vispy\canvas.py", line 33 in __init__
    File "D:\a\napari\napari\napari\_qt\qt_viewer.py", line 410 in _create_canvas
    File "D:\a\napari\napari\napari\_qt\qt_viewer.py", line 224 in __init__
    File "D:\a\napari\napari\napari\_qt\qt_main_window.py", line 112 in __init__
    File "D:\a\napari\napari\napari\_qt\qt_main_window.py", line 565 in __init__
    File "D:\a\napari\napari\napari\viewer.py", line 67 in __init__
    File "D:\a\napari\napari\napari\utils\_testsupport.py", line 242 in actual_factory
    File "D:\a\napari\napari\napari\_qt\_tests\test_qt_provide_theme.py", line 86 in make_napari_viewer_with_plugin_theme
    File "D:\a\napari\napari\napari\_qt\_tests\test_qt_provide_theme.py", line 22 in test_provide_theme_hook_registered_correctly
```
Checked trying to reproduce the test suite failing locally on Windows setting `NAPARI_POPUP_TEST=1` but the test suite pass for me:

```
================= 3457 passed, 41 skipped, 2 xfailed, 9 warnings in 720.14s (0:12:00) =================
```

Also, checking the CI and local test suite final message seems like an error message from `atexit._run_exitfuncs` is being printed like 5 times:

* CI (https://github.com/napari/napari/actions/runs/3960947883/jobs/6785757272#step:9:600):

```
Error in atexit._run_exitfuncs:
  PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp1eehamo7-napari-constraints.txt'
```

* Locally:
```
Error in atexit._run_exitfuncs:
PermissionError: [WinError 32] El proceso no tiene acceso al archivo porque está siendo utilizado por otro proceso: 'C:\\Users\\dalth\\AppData\\Local\\Temp\\tmpm3tu6gaq-napari-constraints.txt'
```

**Comment by dalthviz** [comment]
Note: Re-runs of the bundle workflow jobs for failures to upload artifacts probably due to GitHub Actions degradation status. See https://github.com/actions/upload-artifact/issues/383 and https://www.githubstatus.com/incidents/n18j5g7wpg0s

**Comment by brisvag** [comment]
> Note: seems like the skipped tests are failing due to an access violation error on Windows when Vispy code is creating/managing the OpenGL related context/objects?

Yeah those failures are familiar... Very annoying, sorry about that :/

**Comment by brisvag** [comment]
Before merging, how do others feel about these skipped tests? Should we be more responsible and try to solve the problem instead of shoving it under the rug, or are we happy like this? @Czaki @andy-sweet tagging you since the three of us particularly have fought with these flaky tests before...

**Comment by Czaki** [comment]
I do not understand how these changes may reuires so many changes in tests.

**Comment by andy-sweet** [comment]
> I do not understand how these changes may reuires so many changes in tests.

My assumption is that the skipped tests have underlying issues that are independent of the implementation changes in this PR that mean they fail sometimes. The implementation changes here affect the layout, which in turn affects drawing the widgets, which is something that could be relatively slow on CI runners. My other more speculative guess is that the Windows runners might be a bit slower in this regard and that's the reason we often skip on Windows CI only rather than there being an actual Windows issue.

We already have quite a lot of skips, so I have no issue with the ones added here, especially given the fix is simple and desirable.

**Comment by brisvag** [comment]
Ok, I guess I'll merge then. Let's hope that as these issues keep showing up, we can figure out exactly how to prevent them :)

**Pull request merged**